src/main/scala/com/cloudant/clouseau/IndexService.scala

```diff
@@ -932,4 +914,170 @@ object IndexService {
+      { case (label: RangesLabel, query: RangesQuery) => (label, query) },
+      { case (_container, _element) => throw new ParseException("invalid ranges query") }
+    )
+
+    val asQueriesExtractor = asQueries.Extractor
+
+    val asRange: PartialFunction[Any, List[Product2[RangesName, List[Product2[RangesLabel, RangesQuery]]]]] = ensureElementsType(
+      { case (name: RangesName, asQueriesExtractor(queries)) => (name, queries) },
+      { case (container, _element) => throw new ParseException(container.toString + " is not a valid ranges query") }
+    ).orElse({ case notAList => throw new ParseException("invalid ranges query") })
+
+    val extractor = asRange.Extractor
+
+    getOption[Any](options, 'ranges) match {
+      case Some(extractor(value)) =>
+        Some(value)
+      case None =>
+        None
+    }
+  }
+
+  /**
+   * Returns the Option[List[String]] of a given option key from a map of options if each element of a list is indeed a String.
+   * It also throws ParseException if passed value is not a List[_].
+   *
+   * @param options: Map[Symbol, Any] - map of options to extract given key
+   * @return An optional value of a given key as `Option[List[String]]`
+   * @throws ParseException
+   *
+   */
+  private[clouseau] def getListOfStringsOption(options: Map[Symbol, Any], field: Symbol): Option[List[String]] = {
+    val asListOfStrings: PartialFunction[Any, List[String]] = ensureElementsType(
+      { case string: String => string },
+      { case (container, _element) => throw new ParseException(container.toString + " is not a valid " + field + " query") }
+    ).orElse({ case notAList => throw new ParseException(notAList.toString + " is not a valid " + field + " query") })
+
+    val extractor = asListOfStrings.Extractor
+    getOption[Any](options, field) match {
+      case Some(extractor(value)) =>
+        Some(value)
+      case None =>
+        None
+    }
+  }
+
+  /**
+   * Returns the Option[Set[String]] of a `include_fields` key from given map of options if each element of a list
+   * is indeed a String.
+   *
+   * It does convert List[String] to Set[String] before returning the result. It also injects an "_id" key into Set[String].
+   *
+   * @param options: Map[Symbol, Any] - map of options to extract given key
+   * @return An optional value of a `include_fields` key as `Option[Set[String]]`
+   * @throws ParseException
+   *
+   */
+
+  private[clouseau] def getIncludeFields(options: Map[Symbol, Any]): Option[Set[String]] = {
+    getListOfStringsOption(options, 'include_fields).map(value => Set[String]() ++ ("_id" :: value).toSet)
+  }
+
+  /**
+   * Returns the Option[List[Product2[Option[GroupName], List[Any]]]] of a `groups` key from given map of options
+   * if each element of a list of tuples with arity 2 and first element is a String.
+   * It also throws ParseException if passed value is not a List[_].
```

Comment:
`ParseException` is already hinted in `@throws`.  Maybe you can move the description over there as well?

Reply:
addressed in a12e954